keyboards/planck/rev7/matrix.c

```diff
@@ -0,0 +1,226 @@
+    matrix_scan_user();
+}
+
+void matrix_init(void) {
+    dprintf("matrix init\n");
+    // debug_matrix = true;
+
+    // actual matrix setup - cols
+    for (int i = 0; i < MATRIX_COLS; i++) {
+        setPinOutput(matrix_col_pins[i]);
+    }
+
+    // rows
+    for (int i = 0; i < MATRIX_ROWS; i++) {
+        setPinInputLow(matrix_row_pins[i]);
+    }
+
+    // encoder A & B setup
+    setPinInputLow(B12);
+    setPinInputLow(B13);
+
+    memset(matrix, 0, MATRIX_ROWS * sizeof(matrix_row_t));
+    memset(matrix_debouncing, 0, MATRIX_COLS * sizeof(matrix_row_t));
+
+    matrix_init_kb();
+
+    // setup watchdog timer for 1 second
+    wdgInit();
+
+    static WDGConfig wdgcfg;
+    wdgcfg.pr   = STM32_IWDG_PR_S(1.0);
+    wdgcfg.rlr  = STM32_IWDG_RL_S(1.0);
+    wdgcfg.winr = STM32_IWDG_WIN_DISABLED;
+    wdgStart(&WDGD1, &wdgcfg);
+}
+
+bool encoder_update(uint8_t index, uint8_t state) {
+    bool    changed = false;
+    uint8_t i       = index;
+
+    encoder_pulses[i] += encoder_LUT[state & 0xF];
+
+    if (encoder_pulses[i] >= ENCODER_RESOLUTION) {
+        encoder_value[index]++;
+        changed = true;
+#ifdef ENCODER_MAP_ENABLE
+        encoder_exec_mapping(index, false);
+#else  // ENCODER_MAP_ENABLE
+        encoder_update_kb(index, false);
+#endif // ENCODER_MAP_ENABLE
+    }
+    if (encoder_pulses[i] <= -ENCODER_RESOLUTION) {
+        encoder_value[index]--;
+        changed = true;
+#ifdef ENCODER_MAP_ENABLE
+        encoder_exec_mapping(index, true);
+#else  // ENCODER_MAP_ENABLE
+        encoder_update_kb(index, true);
+#endif // ENCODER_MAP_ENABLE
+    }
+    encoder_pulses[i] %= ENCODER_RESOLUTION;
+#ifdef ENCODER_DEFAULT_POS
+    encoder_pulses[i] = 0;
+#endif
+    return changed;
+}
+
+uint8_t matrix_scan(void) {
+    // reset watchdog
+    wdgReset(&WDGD1);
+
+    // actual matrix
+    for (int col = 0; col < MATRIX_COLS; col++) {
+        matrix_row_t data = 0;
+
+        // strobe col
+        writePinHigh(matrix_col_pins[col]);
+
+        // need wait to settle pin state
+        wait_us(20);
+
+        // read row data
+        for (int row = 0; row < MATRIX_ROWS; row++) {
+            data |= (readPin(matrix_row_pins[row]) << row);
+        }
+
+        // unstrobe col
+        writePinLow(matrix_col_pins[col]);
+
+        if (matrix_debouncing[col] != data) {
+            matrix_debouncing[col] = data;
+            debouncing             = true;
+            debouncing_time        = timer_read();
+        }
+    }
+
+    if (debouncing && timer_elapsed(debouncing_time) > DEBOUNCE) {
+        for (int row = 0; row < MATRIX_ROWS; row++) {
+            matrix[row] = 0;
+            for (int col = 0; col < MATRIX_COLS; col++) {
+                matrix[row] |= ((matrix_debouncing[col] & (1 << row) ? 1 : 0) << col);
+            }
+        }
+        debouncing = false;
+    }
+
+    // encoder-matrix functionality
+
+    // set up C/rows for encoder read
+    for (int i = 0; i < MATRIX_ROWS; i++) {
```

Comment:
so the C/gnd pins are wired to the matrix?

Reply:
The C pins, yeah - this is the same setup used in #7209